### ISSUE TITLE
Cumulative sum (window functions) are much slower than stock pandas

### ISSUE THREAD

**Issue by daviskirk** [duckdb/duckdb#3453]
#### What happens?

Window function to compute cumulative sum is much slower than equivalent pandas code.

Their might be a more efficient way to get the equivalent result out of duck db but I'm using a `sum` with a window to obtain the solution which doesn't seem to be very efficient. 

I'm excited about the possibilities duckdb would give us and we were thinking about replacing some of my pandas code with more declarative SQL but it seems that these "cumulative" queries are still a bottleneck for us (simple group by and aggregations by contrast are super quick in duckdb).

The following is an attempt to show an example, producing the same result with duckdb and pandas:

#### To Reproduce

```
In [1]: import pandas as pd; import numpy as np; import duckdb; c = duckdb.connect()

In [2]: n = 1000_000; m=10_000; df = pd.DataFrame({"a": np.random.random(n), "b": np.random.random(n), "c": np.random.randint(0, m, n)}).reset_index()

In [3]: %timeit c.execute("SELECT index, (SUM(a) over (PARTITION BY c  ORDER BY index ROWS BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW)) AS a FROM df ORDER BY index")
554 ms ± 4.31 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [4]: %timeit df.groupby("c")["a"].cumsum().reset_index()
24 ms ± 123 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

```

Result looks like this (obviously different regarding the random numbers):
```
         index          a
0            0   0.401120
1            1   0.139281
2            2   0.167592
3            3   0.688402
4            4   0.118918
...        ...        ...
999995  999995  56.911257
999996  999996  44.170181
999997  999997  47.532669
999998  999998  50.699650
999999  999999  48.280013
```

#### Environment (please complete the following information):
 - OS: Ubuntu 20.04.4 LTS
 - DuckDB Version: duckdb python package version: 0.3.3
 - DuckDB Client: Python

**Comment by hawkfish** [comment]
I'll have a look to make sure there isn't something dumb, but I suspect that @Alex-Monahan is right. The current operator even uses our fast code and hash partitioning, but doing nothing is always faster than doing something!

**Comment by daviskirk** [comment]
Thanks for the great explanation @Alex-Monahan.  I was just quite surprised by the scale of the difference.  I think in most cases, the better efficiency of more complex queries would offset "special cases" where pandas is faster, but in this case the 10x difference is quite something to overcome.
My use cases also might be special as I'm using cumulative sums and cumulative max in various nested queries so the problems kind of add up, but I understand why the pre-existing order would make a difference.  

If anyone else has smart hacks to somehow deal with cumulative operators (cummax, cumsum and so on) I'd be happy to try them out (converting back and forth between pandas and duckdb is currently an ugly solution but works in come cases).

**Comment by hawkfish** [comment]
I'm measuring about 165ms being spent on the `Fuse` operation and another 78ms copying the data for sorting. So there may be a bit we can do here. But realistically, Pandas can leverage the existing sort order, which is something we have on our roadmap, but may be a while.

**Comment by hawkfish** [comment]
Actually there may be something dumb here - we are making an entire `ChunkCollection` and then dumping it into the sorting logic without ever using it as a collection. Just scanning and building the individual `DataChunk`s should be faster.

**Comment by hawkfish** [comment]
Yes about 30% faster.  

I think we are also copying instead of slicing, which would be another performance sink...

**Comment by eitsupi** [comment]
Hi, I was trying to use `duckdb` to add columns per group, which is not supported by `arrow` R package 8.0.0, and noticed that this is slower than in R dplyr.

```r
set.seed(42)

df <- tidyr::expand_grid(
  date = seq(0, 10^6, by = 1) |> lubridate::as_date(),
  item = 1:4,
  shop = LETTERS[1:4]
) |>
  dplyr::mutate(
    sale = (stats::rnorm(n = dplyr::n()) * 100) |>
      abs() |>
      round(0) |>
      as.integer()
  )

# 1. R only
df |>
  dplyr::group_by(date, item) |>
  dplyr::summarise(sale = sum(sale, na.rm = TRUE), .groups = "drop_last") |>
  dplyr::mutate(total = sum(sale, na.rm = TRUE)) |>
  dplyr::ungroup()

# 2. Arrow -> duckdb
df |>
  arrow::arrow_table() |>
  dplyr::group_by(date, item) |>
  dplyr::summarise(sale = sum(sale, na.rm = TRUE), .groups = "drop_last") |>
  arrow::to_duckdb() |>
  dplyr::mutate(total = sum(sale, na.rm = TRUE)) |>
  dplyr::ungroup() |>
  dplyr::collect()

# 3. Arrow -> R
df |>
  arrow::arrow_table() |>
  dplyr::group_by(date, item) |>
  dplyr::summarise(sale = sum(sale, na.rm = TRUE), .groups = "drop_last") |>
  dplyr::collect() |>
  dplyr::mutate(total = sum(sale, na.rm = TRUE)) |>
  dplyr::ungroup()

# 4. duckdb -> R
df |>
  arrow::arrow_table() |>
  arrow::to_duckdb() |>
  dplyr::group_by(date, item) |>
  dplyr::summarise(sale = sum(sale, na.rm = TRUE), .groups = "drop_last") |>
  dplyr::collect() |>
  dplyr::mutate(total = sum(sale, na.rm = TRUE)) |>
  dplyr::ungroup()
```

My expectation was that the second, "Arrow->duckdb" case would be the fastest, but in fact this case seems noticeably slower than "Arrow->R" or "duckdb->R".

**Comment by hawkfish** [comment]
@eitsupi I must apologise but my understanding of R/dplyr is poor and I want to make sure I understand what the code does. Among other things, I don't see `cumsum` in the example. Is this implied somehow by the use of `sum`?

**Comment by eitsupi** [comment]
@hawkfish Thanks for the reply. Sorry, my explanation was not clear enough.

We can see the actual SQL statement used by executing `dplyr::show_query` at the last line instead of `dplyr::collect`.

```r
set.seed(42)

df <- tidyr::expand_grid(
  date = seq(0, 10^6, by = 1) |> lubridate::as_date(),
  item = 1:4,
  shop = LETTERS[1:4]
) |>
  dplyr::mutate(
    sale = (stats::rnorm(n = dplyr::n()) * 100) |>
      abs() |>
      round(0) |>
      as.integer()
  )

# 2. Arrow -> duckdb
df |>
  arrow::arrow_table() |>
  dplyr::group_by(date, item) |>
  dplyr::summarise(sale = sum(sale, na.rm = TRUE), .groups = "drop_last") |>
  arrow::to_duckdb() |>
  dplyr::mutate(total = sum(sale, na.rm = TRUE)) |>
  dplyr::ungroup() |>
  dplyr::show_query()
```

Output is like this.

```sql
SELECT *, SUM("sale") OVER (PARTITION BY "date") AS "total"
FROM "arrow_001"
```

This process seems to be quite slow.

The table is as follows just before this process is performed.

```txt
# A tibble: 4,000,004 × 3
# Groups:   date [1,000,001]
   date        item  sale
   <date>     <int> <int>
 1 1970-01-01     1   292
 2 1970-01-01     2   211
 3 1970-01-01     3   567
 4 1970-01-01     4   244
 5 1970-01-02     1   670
 6 1970-01-02     2   347
 7 1970-01-02     3   435
 8 1970-01-02     4   226
 9 1970-01-03     1   387
10 1970-01-03     2   408
# … with 3,999,994 more rows
```

Since this is such a row-heavy process, I was quite surprised that duckdb was slower than processing by dplyr.

**Comment by hawkfish** [comment]
Thanks - that help a lot. It looks like there are a large number of small partitions, which is an interesting case. But are the partitions just sorted randomly? It looks like the first partition is sorted 2,4,1,3 for the cumulative sum?

**Comment by eitsupi** [comment]
> But are the partitions just sorted randomly? It looks like the first partition is sorted 2,4,1,3 for the cumulative sum?

The `data` column should be sorted and the `sale` column is not sorted after being generated as a random number (by the `rnorm()` function).

This calculation first tries to calculate the total sales for each day in order to calculate the sales share of each item for each day.
In other words, it is not a cumulative sum (I think), but a calculation of the total value for each group (`date`).

I thought about starting a new issue, but I posted it here because I thought it was relevant to this issue.
Sorry for the confusion.

**Comment by hawkfish** [comment]
> I thought about starting a new issue, but I posted it here because I thought it was relevant to this issue.

This is the same issue so no worries. It looks like we are spending 70% of our time partitioning and sorting the large number of partitions.

**Comment by hawkfish** [comment]
OK, this is something that had been worrying me for a while. The problem is that we have a lot of small partitions, and we hash then into 1024 groups (a la Leis et al.) so we can process them in parallel. But the way we do this, we end up taking 1024 passes over the 4M row input, which is rather dumb.  (This is m code so I can mock it!)

This is good to know because we are designing the out-of-core version, and when to partition was one of my worries, but clearly partitioning early is where we want to be!

**Comment by eitsupi** [comment]
In my use case above, I was able to confirm that duckdb 0.5.0 is very fast.
Thank you for the great library!

**Comment by Alex-Monahan** [comment]
> In my use case above, I was able to confirm that duckdb 0.5.0 is very fast. Thank you for the great library!

If you happen to have some microbenchmark results to share that would be awesome!

**Comment by eitsupi** [comment]
> If you happen to have some microbenchmark results to share that would be awesome!

In this above, `duckdb` is several times faster than `dplyr` only.
For reference, `dtplyr` (`data.table`) is more faster, but we should consider that the data source is an R data frame, so the conditions favor `data.table`.

![image](https://user-images.githubusercontent.com/50911393/188699007-afd5621e-7203-4408-a36a-4876c5057f13.png)

```r
set.seed(42)

df <- tidyr::expand_grid(
  date = seq_len(10^6) |> lubridate::as_date(),
  item = 1:4,
  shop = LETTERS[1:4]
) |>
  dplyr::mutate(
    sale = (stats::rnorm(n = dplyr::n()) * 100) |>
      abs() |>
      round(0) |>
      as.integer()
  )

res <- bench::mark(
  check = FALSE, # Only duckdb returns dbl, others return int.
  min_iterations = 5,
  dplyr = df |>
    dplyr::group_by(date, item) |>
    dplyr::summarise(sale = sum(sale, na.rm = TRUE), .groups = "drop_last") |>
    dplyr::mutate(total = sum(sale, na.rm = TRUE)) |>
    dplyr::ungroup(),
  dtplyr = df |>
    dtplyr::lazy_dt() |>
    dplyr::group_by(date, item) |>
    dplyr::summarise(sale = sum(sale, na.rm = TRUE), .groups = "drop_last") |>
    dplyr::mutate(total = sum(sale, na.rm = TRUE)) |>
    dplyr::ungroup() |>
    dplyr::collect(),
  duckdb = df |>
    arrow::to_duckdb() |>
    dplyr::group_by(date, item) |>
    dplyr::summarise(sale = sum(sale, na.rm = TRUE), .groups = "drop_last") |>
    dplyr::mutate(total = sum(sale, na.rm = TRUE)) |>
    dplyr::ungroup() |>
    dplyr::collect(),
  `duckdb -> dplyr` = df |>
    arrow::to_duckdb() |>
    dplyr::group_by(date, item) |>
    dplyr::summarise(sale = sum(sale, na.rm = TRUE), .groups = "drop_last") |>
    dplyr::collect() |>
    dplyr::mutate(total = sum(sale, na.rm = TRUE)) |>
    dplyr::ungroup()
)

res |> ggplot2::autoplot("violin")
```

**Comment by hawkfish** [comment]
Very cool - thanks!  More coming later this months I hope.

**Comment by hannes** [comment]
I guess we can close this now
